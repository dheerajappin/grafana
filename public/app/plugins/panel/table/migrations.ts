import { PanelModel, FieldMatcherID, ConfigOverrideRule } from '@grafana/data';
import { ReduceTransformerOptions } from '@grafana/data/src/transformations/transformers/reduce';
import { Options } from './types';

/**
 * At 7.0, the `table` panel was swapped from an angular implementation to a react one.
 * The models do not match, so this process will delegate to the old implementation when
 * a saved table configuration exists.
 */
export const tableMigrationHandler = (panel: PanelModel<Options>): Partial<Options> => {
  // Table was saved as an angular table, lets just swap to the 'table-old' panel
  if (!panel.pluginVersion && (panel as any).columns) {
    console.log('Was angular table', panel);
  }

  // Nothing changed
  return panel.options;
};

const transformsMap = {
  timeseries_to_rows: 'seriesToRows',
  timeseries_to_columns: 'seriesToColumns',
  timeseries_aggregations: 'reduce',
};

const columnsMap = {
  avg: 'mean',
  min: 'min',
  max: 'max',
  total: 'sum',
  current: 'last',
  count: 'count',
};

const colorModeMap = {
  cell: 'color-background',
  row: 'color-background',
  value: 'color-text',
};

const migrateTransformations = (panel: PanelModel<Partial<Options>> | any, oldOpts: any) => {
  const transformations = panel.transformations ?? [];
  if (Object.keys(transformsMap).includes(oldOpts.transform)) {
    const opts: ReduceTransformerOptions = {
      reducers: [],
    };
    if (oldOpts.transform === 'timeseries_aggregations') {
      opts.includeTimeField = false;
      opts.reducers = oldOpts.columns.map((column: any) => columnsMap[column.value]);
    }
    transformations.push({
      id: transformsMap[oldOpts.transform],
      options: opts,
    });
  }
  return transformations;
};

const migrateTableStyleToOverride = (style: any) => {
  const fieldMatcherId = /^\/.*\/$/.test(style.pattern) ? FieldMatcherID.byRegexp : FieldMatcherID.byName;
  const override: ConfigOverrideRule = {
    matcher: {
      id: fieldMatcherId,
      options: style.pattern,
    },
    properties: [],
  };

  if (style.alias) {
    override.properties.push({
      id: 'displayName',
      value: style.alias,
    });
  }

  if (style.unit) {
    override.properties.push({
      id: 'unit',
      value: style.unit,
    });
  }

  if (style.decimals) {
    override.properties.push({
      id: 'decimals',
      value: style.decimals,
    });
  }

  if (style.type === 'date') {
    override.properties.push({
      id: 'unit',
      value: `time: ${style.dateFormat}`,
    });
  }

  if (style.colorMode) {
    override.properties.push({
      id: 'custom.displayMode',
      value: colorModeMap[style.colorMode],
    });
  }

  if (style.thresholds?.length) {
    override.properties.push({
      id: 'thresholds',
      value: {
        mode: 'absolute',
        steps: [-Infinity, ...style.thresholds].map((threshold, idx) => ({
          color: style.colors[idx],
          value: parseInt(threshold, 10),
        })),
      },
    });
  }

  return override;
};

/**
 * This is called when the panel changes from another panel
 */
export const tablePanelChangedHandler = (
  panel: PanelModel<Partial<Options>> | any,
  prevPluginId: string,
  prevOptions: any
) => {
  // Changing from angular table panel
  if (prevPluginId === 'table-old' && prevOptions.angular) {
    const oldOpts = prevOptions.angular;
    const transformations = migrateTransformations(panel, oldOpts);
    const overrides = oldOpts.styles.filter((style: any) => style.pattern !== '/.*/').map(migrateTableStyleToOverride);
    panel.transformations = transformations;
    panel.fieldConfig = {
      overrides,
    };
  }

  return {};
};
