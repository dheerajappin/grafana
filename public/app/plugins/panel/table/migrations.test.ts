import { PanelModel } from '@grafana/data';
import { tablePanelChangedHandler } from './migrations';

describe('Table Migrations', () => {
  it('migrates transform out to core transforms', () => {
    const toColumns = {
      angular: {
        columns: [],
        styles: [],
        transform: 'timeseries_to_columns',
        options: {},
      },
    };
    const toRows = {
      angular: {
        columns: [],
        styles: [],
        transform: 'timeseries_to_rows',
        options: {},
      },
    };
    const toAggregations = {
      angular: {
        columns: [
          {
            text: 'Avg',
            value: 'avg',
            $$hashKey: 'object:82',
          },
          {
            text: 'Max',
            value: 'max',
            $$hashKey: 'object:83',
          },
          {
            text: 'Current',
            value: 'current',
            $$hashKey: 'object:84',
          },
        ],
        styles: [],
        transform: 'timeseries_aggregations',
        options: {},
      },
    };

    const columnsPanel = {} as PanelModel;
    tablePanelChangedHandler(columnsPanel, 'table-old', toColumns);
    expect(columnsPanel).toMatchSnapshot();
    const rowsPanel = {} as PanelModel;
    tablePanelChangedHandler(rowsPanel, 'table-old', toRows);
    expect(rowsPanel).toMatchSnapshot();
    const aggregationsPanel = {} as PanelModel;
    tablePanelChangedHandler(aggregationsPanel, 'table-old', toAggregations);
    expect(aggregationsPanel).toMatchSnapshot();
  });

  it('migrates styles to field config overrides', () => {
    const oldStyles = {
      angular: {
        columns: [],
        styles: [
          {
            alias: 'Time',
            align: 'auto',
            dateFormat: 'YYYY-MM-DD HH:mm:ss',
            pattern: 'Time',
            type: 'date',
            $$hashKey: 'object:195',
          },
          {
            alias: '',
            align: 'auto',
            colorMode: 'cell',
            colors: ['rgba(245, 54, 54, 0.9)', 'rgba(237, 129, 40, 0.89)', 'rgba(50, 172, 45, 0.97)'],
            dateFormat: 'YYYY-MM-DD HH:mm:ss',
            decimals: 2,
            mappingType: 1,
            pattern: 'ColorCell',
            thresholds: ['5', '10'],
            type: 'number',
            unit: 'currencyUSD',
            $$hashKey: 'object:196',
          },
          {
            alias: '',
            align: 'auto',
            colorMode: 'value',
            colors: ['rgba(245, 54, 54, 0.9)', 'rgba(237, 129, 40, 0.89)', 'rgba(50, 172, 45, 0.97)'],
            dateFormat: 'YYYY-MM-DD HH:mm:ss',
            decimals: 2,
            mappingType: 1,
            pattern: 'ColorValue',
            thresholds: ['5', '10'],
            type: 'number',
            unit: 'Bps',
            $$hashKey: 'object:197',
          },
          {
            unit: 'short',
            type: 'number',
            alias: '',
            decimals: 2,
            colors: ['rgba(245, 54, 54, 0.9)', 'rgba(237, 129, 40, 0.89)', 'rgba(50, 172, 45, 0.97)'],
            colorMode: null,
            pattern: '/.*/',
            thresholds: [],
            align: 'right',
          },
        ],
      },
    };
    const panel = {} as PanelModel;
    tablePanelChangedHandler(panel, 'table-old', oldStyles);
    expect(panel).toMatchSnapshot();
  });
});
