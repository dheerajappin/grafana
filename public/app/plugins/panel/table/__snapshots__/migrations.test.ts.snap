// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Migrations migrates styles to field config overrides 1`] = `
Object {
  "fieldConfig": Object {
    "overrides": Array [
      Object {
        "matcher": Object {
          "id": "byName",
          "options": "Time",
        },
        "properties": Array [
          Object {
            "id": "displayName",
            "value": "Time",
          },
          Object {
            "id": "unit",
            "value": "time: YYYY-MM-DD HH:mm:ss",
          },
        ],
      },
      Object {
        "matcher": Object {
          "id": "byName",
          "options": "ColorCell",
        },
        "properties": Array [
          Object {
            "id": "unit",
            "value": "currencyUSD",
          },
          Object {
            "id": "decimals",
            "value": 2,
          },
          Object {
            "id": "custom.displayMode",
            "value": "color-background",
          },
          Object {
            "id": "thresholds",
            "value": Object {
              "mode": "absolute",
              "steps": Array [
                Object {
                  "color": "rgba(245, 54, 54, 0.9)",
                  "value": NaN,
                },
                Object {
                  "color": "rgba(237, 129, 40, 0.89)",
                  "value": 5,
                },
                Object {
                  "color": "rgba(50, 172, 45, 0.97)",
                  "value": 10,
                },
              ],
            },
          },
        ],
      },
      Object {
        "matcher": Object {
          "id": "byName",
          "options": "ColorValue",
        },
        "properties": Array [
          Object {
            "id": "unit",
            "value": "Bps",
          },
          Object {
            "id": "decimals",
            "value": 2,
          },
          Object {
            "id": "custom.displayMode",
            "value": "color-text",
          },
          Object {
            "id": "thresholds",
            "value": Object {
              "mode": "absolute",
              "steps": Array [
                Object {
                  "color": "rgba(245, 54, 54, 0.9)",
                  "value": NaN,
                },
                Object {
                  "color": "rgba(237, 129, 40, 0.89)",
                  "value": 5,
                },
                Object {
                  "color": "rgba(50, 172, 45, 0.97)",
                  "value": 10,
                },
              ],
            },
          },
        ],
      },
    ],
  },
  "transformations": Array [],
}
`;

exports[`Table Migrations migrates transform out to core transforms 1`] = `
Object {
  "fieldConfig": Object {
    "overrides": Array [],
  },
  "transformations": Array [
    Object {
      "id": "seriesToColumns",
      "options": Object {
        "reducers": Array [],
      },
    },
  ],
}
`;

exports[`Table Migrations migrates transform out to core transforms 2`] = `
Object {
  "fieldConfig": Object {
    "overrides": Array [],
  },
  "transformations": Array [
    Object {
      "id": "seriesToRows",
      "options": Object {
        "reducers": Array [],
      },
    },
  ],
}
`;

exports[`Table Migrations migrates transform out to core transforms 3`] = `
Object {
  "fieldConfig": Object {
    "overrides": Array [],
  },
  "transformations": Array [
    Object {
      "id": "reduce",
      "options": Object {
        "includeTimeField": false,
        "reducers": Array [
          "mean",
          "max",
          "last",
        ],
      },
    },
  ],
}
`;
